src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,290 @@
+
+  // build the output table. There are two cases:
+  // (1) if the result set is empty (i.e. no rows match
+  // the filter condition) we create an empty table. The empty table
+  // has the same column definitions (i.e. column names and types) as the input table.
+  // (2) if there are rows in the result set we create a new table with the
+  // chunks that we have already constructed and keeping the input table's column
+  // defintions. 
+  // We copy the column definitions from the input table by 
+  // using Table's specialized constructors.
+  if (result_chunks_ptr->empty()) {
+    return std::make_shared<Table>(in_table_ptr);
+  } else {
+    return std::make_shared<Table>(*result_chunks_ptr, in_table_ptr);
+  }
+}
+
+const std::shared_ptr<std::vector<ChunkOffset>> TableScan::scan_chunk(const std::shared_ptr<const Chunk> chunk_ptr,
+                                                                      const ChunkID chunk_id,
+                                                                      const std::string data_type) const {
+  // determine the set of rows that should be included in the scan output,
+  // i.e. find the row indexes that match the filter condition.
+
+  // accumulate the indexes of the rows that should be included in the output in include_rows_ptr.
+  // The container is initialized later on when we perform the actual filtering.
+  // This declaration is needed so that we can pass the object outside of the
+  // scope of the lambda below.
+  std::shared_ptr<std::vector<ChunkOffset>> include_rows_ptr;
+
+  // get segment that we want to filter on and cast it to the right type. Then,
+  // perform a scan on the segment based on the segment type (value/dict/reference).
+  auto segment_ptr = chunk_ptr->get_segment(_column_id);
+  resolve_data_type(data_type, [&](auto type) {
+    using Type = typename decltype(type)::type;
+    // case 1: segment is value segment
+    const auto typed_value_segment_ptr = std::dynamic_pointer_cast<ValueSegment<Type>>(segment_ptr);
+    if (typed_value_segment_ptr) {
+      include_rows_ptr = scan_segment<Type>(typed_value_segment_ptr);
+      return;
+    }
+    // case 2: segment is dictionary segment
+    const auto typed_dict_segment_ptr = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment_ptr);
+    if (typed_dict_segment_ptr) {
+      include_rows_ptr = scan_segment<Type>(typed_dict_segment_ptr);
+      return;
+    }
+    // case 3: segment is reference segment
+    const auto ref_segment_ptr = std::dynamic_pointer_cast<ReferenceSegment>(segment_ptr);
+    if (ref_segment_ptr) {
+      include_rows_ptr = scan_segment<Type>(ref_segment_ptr);
+      return;
+    }
+
+    // we do not support any segment types beyond value, dict, and reference segments
+    throw std::runtime_error("unrecognized segment class at chunk id " + std::to_string(chunk_id) + " and column id " +
+                             std::to_string(_column_id));
+  });
+
+  include_rows_ptr->shrink_to_fit();
```

Comment:
If I am not mistaken you haven't presized the vector. So this will not have an effect.